### PR TITLE
feat: add samples

### PR DESCRIPTION
**Pull request by HemangChothani**

### CODE REVIEW

File: samples/autocommit.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+    conn.execution_options(isolation_level="AUTOCOMMIT")
+    print("Connection autocommit mode is {}".format(level))
+    print("Spanner DBAPI autocommit mode is {}".format(conn.connection.connection.autocommit))
+
+    # [END sqlalchemy_spanner_autocmmit_on]
+    return conn
+
+
+if __name__ == "__main__":
```
Comment (larkee) [inline samples/autocommit.py:30]: @skuruppu Are we moving away from running samples from the command line in all languages?

Comment (skuruppu) [reply samples/autocommit.py:30]: The new style guides are consist for all languages and they recommend not having a CLI for samples.

File: samples/autocommit.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+# Use of this source code is governed by a BSD-style
+# license that can be found in the LICENSE file or at
+# https://developers.google.com/open-source/licenses/bsd
+
+from sqlalchemy import create_engine
+
+
+def enable_autocommit_mode(url):
+    """Enables autocommit mode."""
+    # [START sqlalchemy_spanner_autocmmit_on]
```
Comment (vi3k6i5) [inline samples/autocommit.py:12]: nit: autoc"o"mmit

File: samples/autocommit.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    level = conn.get_isolation_level()
+
+    print("Connection autocommit default mode is {}".format(level))
+    print("Spanner DBAPI default autocommit mode is {}".format(conn.connection.connection.autocommit))
+
+    conn.execution_options(isolation_level="AUTOCOMMIT")
+    print("Connection autocommit mode is {}".format(level))
+    print("Spanner DBAPI autocommit mode is {}".format(conn.connection.connection.autocommit))
+
+    # [END sqlalchemy_spanner_autocmmit_on]
```
Comment (vi3k6i5) [inline samples/autocommit.py:24]: nit: autocommit

File: samples/table_fetch_rows.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
+# Copyright 2021 Google LLC
+#
+# Use of this source code is governed by a BSD-style
+# license that can be found in the LICENSE file or at
+# https://developers.google.com/open-source/licenses/bsd
+
+
+def fetch_rows(table):
+    """fetch all rows from the table"""
```
Comment (vi3k6i5) [inline samples/table_fetch_rows.py:9]: nit: Fetch

File: samples/table_fetch_rows_with_order_by.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
+# Copyright 2021 Google LLC
+#
+# Use of this source code is governed by a BSD-style
+# license that can be found in the LICENSE file or at
+# https://developers.google.com/open-source/licenses/bsd
+
+
+def fetch_rows_with_order_by(table):
+    """fetch all rows from the table in order"""
```
Comment (vi3k6i5) [inline samples/table_fetch_rows_with_order_by.py:9]: nit: Fetch

File: samples/table_filter_data_endswith.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
+# Copyright 2021 Google LLC
+#
+# Use of this source code is governed by a BSD-style
+# license that can be found in the LICENSE file or at
+# https://developers.google.com/open-source/licenses/bsd
+
+
+def filter_data_endswith(table):
+    """filter data with endswith from the table"""
```
Comment (vi3k6i5) [inline samples/table_filter_data_endswith.py:9]: nit:Filter

File: samples/table_filter_data_startswith.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
+# Copyright 2021 Google LLC
+#
+# Use of this source code is governed by a BSD-style
+# license that can be found in the LICENSE file or at
+# https://developers.google.com/open-source/licenses/bsd
+
+
+def filter_data_startswith(table):
+    """filter data with startswith from the table"""
```
Comment (vi3k6i5) [inline samples/table_filter_data_startswith.py:9]: nit:Filter

File: samples/table_filter_data_with_contains.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
+# Copyright 2021 Google LLC
+#
+# Use of this source code is governed by a BSD-style
+# license that can be found in the LICENSE file or at
+# https://developers.google.com/open-source/licenses/bsd
+
+
+def filter_data_with_contains(table):
+    """filter data with contains from the table"""
```
Comment (vi3k6i5) [inline samples/table_filter_data_with_contains.py:9]: nit:Filter

File: samples/table_filter_data_with_like.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
+# Copyright 2021 Google LLC
+#
+# Use of this source code is governed by a BSD-style
+# license that can be found in the LICENSE file or at
+# https://developers.google.com/open-source/licenses/bsd
+
+
+def filter_data_with_like(table):
+    """filter data with like from the table"""
```
Comment (vi3k6i5) [inline samples/table_filter_data_with_like.py:9]: nit: Filter

**Review by vi3k6i5 — APPROVED** [review]
LGTM except some nits. Wait for @larkee to approve before submitting.

File: samples/table_fetch_rows_with_limit.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    # TODO(developer): Create the table
+    # table = Table(
+    #    table_name,
+    #    metadata,
+    #    Column("user_id", Integer, primary_key=True),
+    #    Column("user_name", String(16), nullable=False),
+    # )
+    # table.create()
+
+    table.insert().execute([
```
Comment (larkee) [inline samples/table_fetch_rows_with_limit.py:20]: This should not be here; this belongs in the test.

The sample functions should be contained within the sample tags, i.e.
```
# [START sample_tag]
def function(arg):
    ...
# [END sample_tag]
```

This means there shouldn't be any return values.

File: samples/create_table.py [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+        Integer,
+        MetaData,
+        String,
+        Table,
+        create_engine,
+)
+
+
+def create_table(url, random_table_id):
+    """Create the table"""
```
Comment (larkee) [inline samples/create_table.py:18]: Please ensure all docstrings end with a full stop:
```suggestion
    """Create the table."""
```

**Review by larkee — COMMENTED** [review]
Please move all the tests into a single test file.

Also, please add a nox session to run the sample tests.

File: samples/snippets.py [hunk 12]
```diff
@@ -0,0 +1,479 @@
...
+    metadata = MetaData(bind=engine)
+
+    table = Table(
+        random_table_id,
+        metadata,
+        Column("user_id", Integer, primary_key=True),
+        Column("user_name", String(16), nullable=False),
+    )
+
+    table.create()
```
Comment (larkee) [inline samples/snippets.py:101]: This sample is for dropping a table. There is no reason to include the table creation here. That should be done in the test.

So the test should create the table and then run the sample.

File: samples/snippets.py [hunk 13]
```diff
@@ -0,0 +1,479 @@
...
+def delete_all_rows(table):
+    """Delete all rows from the table.
+
+    The table must already exist and can be created using
+    `create_table.`
+    """
+    insert_data(table, DATA_2)
+
+    result = table.select().execute().fetchall()
+    print("Total inserted rows:", len(result))
```
Comment (larkee) [inline samples/snippets.py:154]: Inserting the data should be part of the test, not the sample. Also applies to the below samples (except for `insert_row` sample).

**Comment by skuruppu** [comment]
@IlyaFaer can you please take over this PR and address the outstanding feedback? I think it's a good time to start merging in this PR.

**Comment by skuruppu** [comment]
I think we also need to fix the snippet-bot check. My guess is that it wants the snippet names to start with `spanner_`.

**Comment by IlyaFaer** [comment]
@googlebot I consent

**Comment by IlyaFaer** [comment]
@skuruppu, @larkee, okay, I think most of the errors fixed. Let's take another look.

**Comment by skuruppu** [comment]
@larkee if you can please review this again, that would be great. We need this before we can cut a 1.0 release.

File: samples/snippets_test.py [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+
+    for col in table.columns:
+        assert col.name in out
+
+
+def test_table_get_foreign_key(capsys, db_url, table_w_foreign_key):
+    snippets.get_table_foreign_key(db_url, table_w_foreign_key)
+    out, err = capsys.readouterr()
+
+    assert "Fetched foreign key:" in out
```
Comment (larkee) [inline samples/snippets_test.py:190]:
```suggestion
    assert "Fetched foreign keys:" in out
```

File: samples/snippets_test.py [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+    assert "Rows exist: 6" in out
+    assert "Rows exist after deletion: 0" in out
+
+
+def test_table_delete_row_with_where_clause(capsys, connection, table):
+    insert_data(connection, table, DATA)
+    snippets.delete_row_with_where_clause(connection, table)
+
+    out, err = capsys.readouterr()
+    assert "Rows exist: 6" in out
```
Comment (larkee) [inline samples/snippets_test.py:109]:
```suggestion
    assert "Row count: 6" in out
```

File: samples/snippets_test.py [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+    assert "Rows exist after deletion: 5" in out
```
Comment (larkee) [inline samples/snippets_test.py:110]:
```suggestion
    assert "Row count after deletion: 5" in out
```

File: samples/snippets_test.py [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+    assert "Index created" in out
+    assert indexes[0]["unique"] is True
+
+
+def test_table_delete_all_rows(capsys, connection, table):
+    insert_data(connection, table, DATA)
+    snippets.delete_all_rows(connection, table)
+
+    out, err = capsys.readouterr()
+    assert "Rows exist: 6" in out
```
Comment (larkee) [inline samples/snippets_test.py:100]:
```suggestion
    assert "Row count: 6" in out
```

File: samples/snippets_test.py [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+    assert "Rows exist after deletion: 0" in out
```
Comment (larkee) [inline samples/snippets_test.py:101]:
```suggestion
    assert "Row count after deletion: 0" in out
```

**Review by larkee — COMMENTED** [review]
A bunch of minor nit fixes

**Review by larkee — APPROVED** [review]

**Pull request merged**